api/src/OpenTelemetry/Trace/Core.hs

```diff
@@ -665,6 +680,19 @@ globalTracer = unsafePerformIO $ do
   newIORef p
 {-# NOINLINE globalTracer #-}
 
+-- | This type class allows library authors and users to enrich the
+-- contents of an exception with additional attributes.
+class Exception e => HasExceptionEnricher e where
+  exceptionEnricher :: Proxy e -> ExceptionEnricher
```

Comment:
Hmm. Thinking about how I'd write this for `annotated-exception`.

```haskell
instance (HasExceptionEnricher e) => HasExceptionEnricher (AnnotatedException e) where
  exceptionEnricher _ = ExceptionEnricher
    { exceptionEnricherAdditionalFields = \(AnnotatedException anns er) ->
        exceptionEnricherAdditionalFields (exceptionEnricher (pure er)) e
        <> makeAdditionalFields anns
    , exceptionEnricherName = "AnnotatedException " <> show (typeRep @e)
    }
```

`makeAdditionalFields` is pretty limited- it can only access the `Annotation`, which can only really be `show`n into a `String`. So the only real impl would be something like `map (\(Annotation ann) -> (Text.pack (show $ typeRep (pure ann)), show ann)) annotations`.

I would expect that apps would want to customize this directly, for their own app. So having a type class may not be the best way to do that.

`discoverInstances` also only works "transitively" in the same package, otherwise you do need the type itself imported in a module (and used!) or it won't discover the instance.

Reply:
Yeah, so that's why I took the approach of providing `tracerProviderOptionsExceptionEnrichers`. 

There are two things I'm trying to solve here:

1. This is the `hs-opentelemetry-api` package, which is intended to be usable by libraries (such as persistent) in such a fashion that they have no effect if you aren't using `hs-opentelemetry-sdk` in your application. I'd like to provide a mechanism for domain-specific exceptions to provide additional attributes to the span, which is why the `ExceptionEnricher` + type class + `discoverExceptionEnricher` mechanism exists. I know it's a bit brittle in that you have to import the right things in the right module, but I'm not sure how else to provide any amount of automatic hookup for library-supplied enrichers.
2. I don't want the type class instance to be required to use the library, which is why any discovered enrichers ultimately have to be set in the `tracerProviderOptionsExceptionEnrichers :: [ExceptionEnricher]` record field. In the case of wanting to do more than rendering annotations to a string, this would presumably be where you'd be able to append a domain-specific `ExceptionEnricher` that can leverage `Typeable` to do more useful annotations.

Reply:
Yeah, I think we do need the function parameter to have a `[Enricher]` parameter - otherwise, we can only deal with the things that are known that that point in time. I ran into this same problem when considering a plain function approach - you can recurse into wrappers just fine, and you can compose `SomeException -> [(Text, Attribute)]` fine with `<>`, but you can't make a wrapper that can see through *other* wrappers defined later, unless you accept the `final :: [Enricher]` as a late bound function argument.

Playing with it this afternoon and I got:

```haskell
newtype Enricher = Enricher ([Enricher] -> SomeException -> [(Text, Attribute)])
    deriving newtype (Semigroup, Monoid)

enrich :: SomeException -> [Enricher] -> [(Text, Attribute)]
enrich exception enrichers = 
    foldMap (\(Enricher k) -> k enrichers exception)

simpleEnricher :: Exception e => (e -> [(Text, Attribute)]) -> Enricher
simpleEnricher k =
    Enricher \_ someException -> 
        foldMap k (fromException someException)

nestedEnricher :: Exception e => ([Enricher] -> e -> [(Text, Attribute)]) -> Enricher
nestedEnricher k =
    Enricher \final exception -> 
        foldMap (k final) (fromException exception) 
```

Each library can define an `Enricher` for a specific exception, then `mconcat` them all into a single `Enricher`.

Figuring out the *exact* right implementation for this is subtle, though. `cast` and `fromException` need to be chosen with care, or you may get into infinite loops. So I think it's important to figure out some basic tests (`SomeException`, `SomeAsyncException`, `SyncExceptionWrapper`, `AsyncExceptionWrapper`, `AnnotatedException`, etc) and ensure we're getting the right behavior.

Figuring out the plugin-UX side of things can happen at a later step. IMO it's fine to say "You need to import the `Enricher` for a given library and plug it in to the settings." Very few libraries work in any other way.